woocommerce/woocommerce-gutenberg-products-block
# Convert validation context to data store

This PR is for moving the validation errors to a `@wordpress/data` store instead of using context

Changes made in the PR:

 - Added reducer, selectors, actions types, and actions for the validation store.
 - Added test cases for the validation store.
 - Found instances of `useValidationContext` and refactored them to use the data store.

### Testing

#### Automated Tests


### WooCommerce Visibility




### Manual Testing

1.
- Add a Single Product Block to a page and choose a product with multiple attributes (Hoodie is a good one from the sample products set)
- Go to the page and without selecting any attributes try to add it to your cart.
- Verify you see errors on each input.
- Select an option and ensure the error message disappears as soon as you do.
- Add to cart and verify it is successful.

2.
- Create a page with the Cart Block on it, ensure the Shipping Calculator is enabled.
- Add items to your cart and visit the Cart Block.
- Use the shipping calculator, leave the free text fields blank, ensure validation errors appear.
- Try to calculate shipping with validation errors present. Ensure the form doesn't submit.
- Fix the validation errors, ensure the form submits and the address updates.

3. 
- Add a coupon to your store.
- Add an item to your cart and go to the Cart Block
- Use the coupon input and enter an invalid coupon, try to apply it.
- Ensure an error appears.
- Type into the coupon box again and ensure the error disappears.
- Type a valid coupon code and ensure it submits and the discount is applied to your cart.

4.
- Add the Checkout Block to a page, on the Shipping Address Block set `require phone number` to true.
- In an incognito window add items to your cart and go to the Checkout Block.
- Immediately press the `Place order` button.
- Ensure the invalid fields in the form are highlighted.
- Fill in your details, enter invalid data in each field (fill and then empty again in all cases) and ensure the ones that require validation show you errors.
- You should see errors like this:
- <img width="633" alt="image" src="https://user-images.githubusercontent.com/5656702/172273145-f670551b-3cbd-4edd-8904-9712a5df9eb4.png">
- While errors are present, try to proceed with checkout. Ensure you cannot proceed by pressing the place order button until the errors are fixed.

5.
- Install the Stripe extension and set it up.
- Go to the Cart Block and enter some invalid card details in the Stripe form.
- Ensure you see validation errors on each field: 
<img width="612" alt="image" src="https://user-images.githubusercontent.com/5656702/172273393-c3cc9ea5-0087-4ecf-823c-9cee1495772e.png">
